hilti/include/rt/types/map.h

```diff
@@ -31,76 +31,240 @@ class Map;
+ *
+ * In particular it guarantees that iterators are either valid, or throw an
+ * exception when accessed.
+ *
+ * If not otherwise specified, we follow the semantics on iterator invalidation
+ * of `std::map` with the caveat that iterators which cannot be dereferenced
+ * can become invalid through mutating `Map` operations. We still validate
+ * invalid dereferencing of such iterators.
+ *
+ *     rt::Map<int, int> map;
+ *     auto it = map.begin(); // Valid iterator which cannot be dereferenced.
+ *
+ *     // Mutating the map invalidates not dereferencable iterators.
+ *     map.insert({1, 1});
+ *
+ *     *it; // Iterator now invalid, throws.
+ *
+ * If not otherwise specified, member functions have the semantics of
+ * `std::map` member functions.
+ * */
 template<typename K, typename V>
-class Map : protected std::map<K, V>, public hilti::rt::detail::iterator::Controllee {
+class Map : protected std::map<K, V> {
 public:
     using M = std::map<K, V>;
-    using C = hilti::rt::detail::iterator::Controllee;
+    using C = std::shared_ptr<Map<K, V>*>;
 
+    C _control = std::make_shared<Map<K, V>*>(this);
+
+    using key_type = typename M::key_type;
     using value_type = typename M::value_type;
 
-    using ConstIterator = typename M::const_iterator;
-    using SafeIterator = typename M::iterator;
+    using iterator = typename map::Iterator<K, V>;
+    using const_iterator = typename map::ConstIterator<K, V>;
 
     Map() = default;
     Map(std::initializer_list<value_type> init) : M(std::move(init)) {}
 
-    /** Returns true if a specific key is part of the set. */
-    bool contains(const K& k) { return this->find(k) != this->end(); }
+    /** Checks whether a key is set in the map.
+     *
+     * @param `k` the key to check for
+     * @return `true` if the key is set in the map
+     */
+    bool contains(const K& k) { return this->find(k) != static_cast<const M&>(*this).end(); }
 
     /**
-     * Returns the value for a given key, with an optional default if not foound.
+     * Attempts to get the value for a key.
      *
      * @param k key to retrieve
-     * @param default_ if given, a defautl to return if *k* is not part of the map.
-     *
-     * @throws `IndexError` if `k` is not part of the map and no default has
-     * been given.
-     * FIXME(bbannier): adjust docs.
+     * @return either the value or a nullopt
      */
     std::optional<std::reference_wrapper<const V>> get(const K& k) const {
-        if ( const auto& i = this->find(k); i != this->end() )
+        if ( const auto& i = this->find(k); i != static_cast<const M&>(*this).end() )
             return i->second;
 
         return std::nullopt;
     }
 
-    const V& operator[](const K& k) const { return static_cast<const M&>(*this)[k]; }
-    V& operator[](const K& k) { return static_cast<M&>(*this)[k]; }
+    /** Access an element by key
+     *
+     * This function invalidates all iterators into the map.
```

Comment:
Hmm ... I'm realizing this is actually not great in terms of semantics. First, we can't guarantee that the map will actually be written to, right? So iterators could be invalidated even if no change happens. And second, for an existing key `x`, when just updating its value (e.g., `m["x"] = 42`), it would be pretty surprising to have all iterators be invalidated. Seems that's pretty coming thing to do during a loop over a map for example.

Reply:
Let's make this return an `AssignProxy` which should only invalidate iterators if it actually (lazyily) inserts a new element.

Reply:
A slightly rewritten `AssignProxy` is back now.